### ISSUE TITLE
Crashes on Wayland

### ISSUE THREAD

**Issue by bk138** [LibVNC/x11vnc#5]
...because XGetImage() fails. Either work around it, show a warning or implement the functionality. Unfortunately, grabbing screen contents is not supported (yet) on Wayland. There is a weston-specific screen shooter https://github.com/krh/weston/blob/master/src/screenshooter.c but no xdg-shell API that's supported by all compositors. Also see http://lists.freedesktop.org/archives/wayland-devel/2013-December/012571.html

**Comment by bk138** [comment]
https://github.com/GNOME/gnome-screenshot/blob/master/src/screenshot-utils.c uses DBUS and works on Wayland...

**Comment by bk138** [comment]
Proof of concept http://aleksandarjovanov.blogspot.de/2012/10/having-fun-with-gnome-shell.html

**Comment by bk138** [comment]
https://mail.gnome.org/archives/commits-list/2013-January/msg00410.html

**Comment by Hello71** [comment]
this can be implemented using the "fullscreen shell" protocol. documentation seems poor at this point however.

**Comment by bk138** [comment]
@Hello71 do you have a link?

**Comment by bk138** [comment]
Fixed with 5843f89923a92bbf5e9fb7d7797bb27907590572 and 528baf729f300de12c963e09d509d29d4f44ec1f.

**Comment by jeff-dagenais** [comment]
Hi all,

Is since x11vnc resorts to rawfb when a wayland compositor is detected, is it possible to get a wayland session using a drm/kms backend to also draw on the /dev/fb0 (for example), so that x11vnc can serve this frame buffer using rawfb?

**Comment by jeff-dagenais** [comment]
My end goal here is to run our apps on a wayland compositor (say weston or a QtWayland based compositor) on this HW https://xilinx-wiki.atlassian.net/wiki/spaces/A/pages/18841928/Xilinx+MALI+driver#XilinxMALIdriver-Wayland/GBMbackend: but still use my beloved x11vnc. I am under the impression using gnome-remote-desktop and mutter would be overkill for an embedded system distro (Yocto made).

**Comment by bk138** [comment]
@jeff-dagenais you somehow have to hook into the compositor to get the framebuffer data, very much as 528baf729f300de12c963e09d509d29d4f44ec1f does it in a scripted, not very performant way.

**Comment by jeff-dagenais** [comment]
I'm weary that it has not made it into x11vnc... is it because there are major hurdles in this or it simply has not made it into a priority for any contributor?

**Comment by bk138** [comment]
Both, I'd say. Pull requests welome :-)

**Comment by jeff-dagenais** [comment]
I'm contemplating adding some of this code https://github.com/wayland-project/weston/blob/master/clients/screenshot.c along with the protocol sources (.h and .c) into x11vnc. This obviously would be very experimental and perhaps make a lot of assumptions. It could probably evolve into something useable by many.

My idea is to do the setup parts only onces, which ultimately creates a shm file  which is mmap'd into memory. Then when x11vnc is ready for more data (a refresh), it can refresh the shm data file using the `weston_screenshooter_shoot` API.

I am browsing the x11vnc code now and am having difficulty figuring out where I need to add these little bits. I think my scenario is a `-rawfb weston-screenshooter:1.0` kind of deal ('1' would be the wayland instance, and '0' the output index). The screenshot.c wayland code can figure out all the output size and color format and all. What I don't quite know yet is how to capture the event where x11vnc is ready for a refresh before it looks at the raw fb (mmap'd). I don't think I want to use the `-snapfb` because it creates an extra unwanted copy of the data. Perhaps I can add the call `weston_screenshooter_shoot` right before `scan_for_updates(0);` [here](https://github.com/LibVNC/x11vnc/blob/a556dc7bc141988a1107c48942bace17d2f072a5/src/screen.c#L4671) and have the shm `mmap` call put the mapped address in `raw_fb_addr`?

Thanks for any pointers you may have.

**Comment by bk138** [comment]
Hi Jeff!
Thanks for tackling this! Some thoughts:

* For an MVP  and to gain to most speedup with a minimum amount of work, I suggest augmenting the existing `deskshot` utility with a native version that either provides a shm segment or file to be used by `-rawfb shm:` or `-rawfb map:`.
* For a later, more refined version, it's either another `-rawfb` "protocol" (though generelly no input in this mode besides `-pipeinput`) or something along the lines of the OSX-supporting code with extra code paths for getting the display contents _and_ injecting input.

With lean and iterative design in mind, I'd tackle the MVP first and measure whether the gains are already enough.

**Comment by jeff-dagenais** [comment]
Thanks for your input.

I actually got it working in about 2 hours. Did mostly what I mentioned. The image is good but the performance hit on the weston compositor is crippling. It basically stalls or locks it such that test clients like weston-simple-egl show significant loss in FPS.

Learned a lot doing it. I will keep digging into this. For future reader benefit, I found the lipstick compositor (Mer project) has a "recorder" which seems to work with lipstick2vnc. Looks like it may work for me but mostly demonstrates how relatively simple it is. Don't know about the performance though.

So I am currently looking at the screen-share.c file and how it establishes a link and a `wl_shm` with a second weston instance that runs the rdp-backend.so. The `command` that screen-share.c invokes is configurable (invoke x11vnc instead?). But screen-share.c has expectations of things that should appear in the `wl_registry` (details fuzzy at this time) to begin collaboration with the second weston instance.

I may just start from weston's screen-share.c and create a sibling file which would be based on `rfb/rfb.h` directly. That would cut x11vnc out of my project entirely though (sorry 😕).

In this spirit, I might stop bothering you here as this topic is somewhat not even related to this original issue. Thanks for indulging me though! Cheers!
